abi.md

```diff
@@ -0,0 +1,164 @@
+
+Vectors shall be lowered to Verilog packed vectors.
+
+Bundles shall be recursively split as per "Aggregate Type Lowering", except
+instead of recursively converting bundles to ground types, the recursion stops
+at passive types.
+
+Passive bundles shall be lowered to Verilog packed structs.
+
+Reference types in ports shall be logically split out from aggregates and named
+as though "Aggregate Type Lowering" was used.
+
+## On Types
+
+Types are only guaranteed to follow this lowering when the Verilog type is on an
+element which is part of the ABI defined public elements.  These include
+use in ports and elements exported by reference through a public module.
+
+Ground types are lowered to the `logic`{.verilog} data type (SV.6.3.1), which is
+the 4-valued type.  It is important to distinguish the `logic`{.verilog} data
+type from using `logic`{.verilog} as a keyword to declare a variable instead of
+net-list object.
+
+Both unsigned and signed integers produce an unsigned packed bit vector of the 
+form `[width-1:0]`.  Whether Verilog variable or netlist depends on the 
+construct being used.
+
+Passive bundles, when lowered, are lowered to packed structs with their fields
+recursively following these lowering rules.
+
+Vectors, when lowered, are lowered to packed vectors with their element type
+recursively following these rules.
+
+Enums shall have their payloads lowered as per this section.  By construction,
+enums are passive FIRRTL types, so any valid variant payload will lower to a 
+verilog type.  An enum with empty or 0-bit payloads for all variants will lower 
+to a Verilog enum with tags for each FIRRTL enum tag value.  A FIRRTL enum with
+at least one payload will lower to a packed struct containing a tag field which 
+is a Verilog enum, as well as a data field containing a packed union of the 
+padded types of the payloads.  A padded payload is a packed struct with the 
+payload as the first field and a packed bit vector as a second field.  The 
+padding for each payload is set to ensure all padded payloads have the same bit
+width as required by Verilog packed unions.
+
+# Versioning Scheme of this Document
+
+This is the versioning scheme that applies to version 1.0.0 and later.
+
+The versioning scheme complies with
+[Semantic Versioning 2.0.0](https://semver.org/#semantic-versioning-200).
+
+Specifically,
+
+The PATCH digit is bumped upon release which only includes non-functional changes,
+such as grammar edits, further examples, and clarifications.
+
+The MINOR digit is bumped for feature additions to the spec.
+
+The MAJOR digit is bumped for backwards-incompatible changes such as features
+being removed from the spec, changing their interpretation, or new required
+features being added to the specification.
+
+In other words, any verilog generated that was emitted with `x.y.z` will be 
+compliant with `x.Y.Z`, where `Y >= y`, `z` and `Z` can be any number.
```

Comment:
This can just say that this document is co-versioned with the FIRRTL specification.